src/image.tsx

```diff
@@ -178,55 +178,72 @@ export const ImageContent = React.forwardRef<HTMLImageElement, ImageProps>(
-                            ? undefined
-                            : vars.borderRadii.container,
-                    })
-                )}
-                alt={alt}
-                onError={() => {
-                    setIsError(true);
-                    setIsLoading(false);
-                    setHideLoadingFallback(true);
-                    onError?.();
-                }}
-                onLoad={onLoadHandler}
-            />
-        );
-
         React.useEffect(() => {
             // Needed because there is some race condition with SSR and onLoad events
+            // load event could be fired before the component is hydrated and mounted client side
             // https://github.com/facebook/react/issues/15446
             if (imageRef.current?.complete) {
                 onLoadHandler();
             }
         }, [onLoadHandler]);
 
+        const isLoading =
+            isServerSide() || !(document.getElementById(imageId) as HTMLImageElement | null)?.complete;
+
+        const img = (
+            <>
+                {/* https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/issues/309 */}
+                {/* eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions */}
+                <img
+                    suppressHydrationWarning
+                    id={imageId}
+                    style={{
+                        opacity: isLoading && withLoadingFallback ? 0 : 1,
+                    }}
+                    ref={combineRefs(imageRef, ref)}
+                    src={src}
+                    className={classnames(
+                        styles.image,
+                        {[styles.imageWithBorder]: props.border},
+                        sprinkles({
+                            position: ratio !== 0 ? 'absolute' : 'static',
+                            borderRadius: props.circular
+                                ? '50%'
+                                : noBorderSetting
+                                ? undefined
+                                : vars.borderRadii.container,
+                        })
+                    )}
+                    alt={alt}
+                    onError={() => {
+                        setIsError(true);
+                        setHideLoadingFallback(true);
+                        onError?.();
+                    }}
+                    onLoad={onLoadHandler}
+                />
+                {/* When using SSR, we render a small script that makes the img visible as soon as it finishes loading, without waiting for React client hydrate */}
+                {withLoadingFallback && (
+                    <script
+                        // eslint-disable-next-line react/no-danger
+                        dangerouslySetInnerHTML={{
+                            __html: `document.getElementById("${imageId}").addEventListener('load', (e) => e.target.style.opacity = "1")`,
+                        }}
+                    />
+                )}
```

Comment:
this is the trick to be able to remove the skeleton before hydration

Reply:
clever